pkg/auth/authenticator/password/ldappassword/ldap.go

```diff
@@ -0,0 +1,227 @@
+	// BindPassword is the optional password to bind to for the search phase.
+	BindPassword string
+
+	// AttributeEmail is the optional list of LDAP attributes to use for the email address of the user identity.
+	// The first attribute with a non-empty value is used.
+	AttributeEmail []string
+	// AttributeName is the optional list of LDAP attributes to use for the display name of the user identity.
+	// The first attribute with a non-empty value is used.
+	AttributeName []string
+	// AttributePreferredUsername is the optional list of LDAP attributes to use for the preferred username of the user identity.
+	// The first attribute with a non-empty value is used. If not specified, the id determined by AttributeID is used as the preferred login.
+	AttributePreferredUsername []string
+	// AttributeID is the required list of LDAP attributes to use for the id address of the user identity.
+	// The first attribute with a non-empty value is used. If no attributes have values, login fails.
+	AttributeID []string
+}
+
+// Authenticator validates username/passwords against an LDAP v3 server
+type Authenticator struct {
+	providerName string
+	options      Options
+	mapper       authapi.UserIdentityMapper
+}
+
+// New returns an authenticator which will validate usernames/passwords using LDAP.
+func New(providerName string, options Options, mapper authapi.UserIdentityMapper) (authenticator.Password, error) {
+	auth := &Authenticator{
+		providerName: providerName,
+		options:      options,
+		mapper:       mapper,
+	}
+	return auth, nil
+}
+
+// AuthenticatePassword validates the given username and password against an LDAP server
+func (a *Authenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	identity, ok, err := a.getIdentity(username, password)
+	if err != nil {
+		return nil, false, err
+	}
+	if !ok {
+		return nil, false, nil
+	}
+
+	user, err := a.mapper.UserFor(identity)
+	glog.V(4).Infof("Got userIdentityMapping: %#v", user)
+	if err != nil {
+		return nil, false, fmt.Errorf("Error creating or updating mapping for: %#v due to %v", identity, err)
+	}
+
+	return user, true, nil
+
+}
+
+// getIdentity looks up a username in an LDAP server, and attempts to bind to the user's DN using the provided password
+func (a *Authenticator) getIdentity(username, password string) (authapi.UserIdentityInfo, bool, error) {
+	if len(username) == 0 || len(password) == 0 {
+		return nil, false, nil
+	}
+
+	// Make the connection
+	var (
+		l       *ldap.Conn
+		err     error
+		secured bool
+	)
+	switch a.options.URL.Scheme {
+	case ProtocolLDAP:
+		l, err = ldap.Dial("tcp", a.options.URL.Host)
+	case ProtocolLDAPS:
+		l, err = ldap.DialTLS("tcp", a.options.URL.Host, a.options.TLSConfig)
+		secured = true
+	default:
+		err = fmt.Errorf("unsupported scheme %q", a.options.URL.Scheme)
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	defer l.Close()
+
+	// Upgrade to TLS if needed
+	if !secured && !a.options.Insecure {
```

Comment:
refactored to a `connect` function... the only tricky bit is if the TLS upgrade fails, we have to close the connection ourselves